src/modules/pmc/src/lib/PmcBase.cpp

```diff
@@ -684,29 +685,171 @@ int PmcBase::ConfigureSources(const std::map<std::string, std::string> sources)
+                const auto &key = gpgKeys.find(gpgKeyFileId);
+                if (key != gpgKeys.end())
+                {
+                    const std::string signedByPrefix = "signed-by=";
+                    std::string placeholder = signedByPrefix + gpgKeyFileId;
+                    std::size_t index = packageSource.find(placeholder);
+
+                    if (index != std::string::npos)
+                    {
+                        std::string gpgKeyConfiguration = signedByPrefix + GenerateGpgKeyPath(gpgKeyFileId);
+                        packageSource.replace(index, placeholder.length(), gpgKeyConfiguration);
+                    }
+                }
+            }
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+std::string PmcBase::GenerateGpgKeyPath(const std::string& gpgKeyId)
+{
+    return g_keysFolderPath + gpgKeyId + ".gpg";
+}
+
+int PmcBase::DownloadGpgKeys(const std::map<std::string, std::string>& gpgKeys)
+{
+    int status = PMC_0K;
+
+    for (auto& key : gpgKeys)
+    {
+        m_executionState.SetExecutionState(StateComponent::Running, SubStateComponent::DownloadingGpgKeys, key.first);
+        std::string keyFilePath = GenerateGpgKeyPath(key.first);
+        const std::string& sourceUrl = key.second;
+
+        // Delete file when provided map value is empty
+        if (sourceUrl.empty())
+        {
+            if (FileExists(keyFilePath.c_str()))
+            {
+                if (remove(keyFilePath.c_str()))
+                {
+                    status = errno;
+                    OsConfigLogError(PmcLog::Get(), "Failed to delete key file %s", keyFilePath.c_str());
+                    m_executionState.SetExecutionState(StateComponent::Failed, SubStateComponent::ModifyingSources, key.first);
+                }
+            }
+            else if (IsFullLoggingEnabled())
+            {
+                OsConfigLogInfo(PmcLog::Get(), "Nothing to delete. Key file %s does not exist", keyFilePath.c_str());
+            }
+        }
+        else
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogInfo(PmcLog::Get(), "Downloading GPG key from %s to %s", sourceUrl.c_str(), keyFilePath.c_str());
+            }
+
+            std::string command = std::regex_replace(g_commandDownloadGpgKey, std::regex("\\$url"), sourceUrl);
+            command = std::regex_replace(command, std::regex("\\$destination"), keyFilePath);
+            status = RunCommand(command.c_str(), nullptr);
+
+            if (status != PMC_0K)
+            {
+                OsConfigLogError(PmcLog::Get(), "Failed to download key from %s to %s", sourceUrl.c_str(), keyFilePath.c_str());
+                m_executionState.SetExecutionState(StateComponent::Failed, SubStateComponent::DownloadingGpgKeys, key.first );
+            }
+        }
+    }
+
+    return status;
+}
+
+int PmcBase::DeserializeGpgKeys(const rapidjson::Document& document, DesiredState& state)
```

Comment:
Nit: I would put that function close to the other Deserialize function

Reply:
done